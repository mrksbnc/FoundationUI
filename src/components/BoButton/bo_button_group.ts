export enum BoButtonGroupSize {
	small = 'small',
	default = 'default',
	large = 'large',
}

export enum BoButtonGroupVariant {
	default = 'default',
	dark = 'dark',
	info = 'info',
	success = 'success',
	warning = 'warning',
	danger = 'danger',
}

export enum ButtonGroupEvent {
	/**
	 * @event select
	 * @description Emitted when one of the buttons in the group is clicked
	 * @emits `{ id: string | number, item: BoButtonGroupItem }` The id and the item of the selected button
	 */
	SELECT = 'select',
}
