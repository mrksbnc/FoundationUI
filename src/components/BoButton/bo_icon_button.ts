export enum BoIconButtonType {
	default = 'default',
	round = 'round',
}
