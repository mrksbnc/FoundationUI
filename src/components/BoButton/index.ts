/**
 * @module BoButton
 * @since 0.0.1
 *
 * @description module export for the BoButton component and its members
 */
export { BoButtonType, BoButtonVariant, useBoButton } from './ts/bo_button';
export type {
	BoButtonClassMap,
	UseBoButtonCompoableArgs,
} from './ts/bo_button.types';
export { default as BoButton } from './vue/bo_button.vue';
