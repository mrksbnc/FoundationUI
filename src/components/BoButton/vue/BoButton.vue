<template>
	<button
		type="button"
		role="button"
		:disabled="disabled"
		:class="[style, `bo-button--${variant} bo-button--${type}`]"
		class="flex items-center content-center justify-center space-x-2 shadow-sd focus:ring-1 focus:ring-opacity-50 focus:outline-none transition-all duration-200 ease-in-out"
	>
		<span
			class="bo-button__prefix"
			v-if="prefixIcon"
		>
			<bo-icon
				:icon="prefixIcon"
				:size="BoSize.default"
			/>
		</span>
		<span class="bo-button__label">
			{{ label }}
		</span>
		<span
			class="bo-button__suffix"
			v-if="suffixIcon"
		>
			<bo-icon
				:icon="suffixIcon"
				:size="BoSize.default"
			/>
		</span>
	</button>
</template>

<script setup lang="ts">
import { computed, toRefs, type PropType } from 'vue';
import { BoButtonType, BoIcon, type Icon } from '@/components';
import { BoSize, BoVariant } from '@/constants';
import { TailwindUtils } from '@/utils';

const props = defineProps({
	label: {
		type: String,
		required: true,
	},
	variant: {
		type: String as PropType<BoVariant>,
		default: BoVariant.primary,
	},
	type: {
		type: String as PropType<BoButtonType>,
		default: BoButtonType.default,
	},
	disabled: {
		type: Boolean,
		default: false,
	},
	size: {
		type: String as PropType<BoSize>,
		default: BoSize.default,
	},
	prefixIcon: {
		type: String as PropType<Icon>,
		default: null,
	},
	suffixIcon: {
		type: String as PropType<Icon>,
		default: null,
	},
});

const { label, variant, type, disabled, size, prefixIcon, suffixIcon } =
	toRefs(props);

const disabledClasses = 'disabled:cursor-not-allowed disabled:opacity-50';

const textSizes = computed<string>(() => {
	switch (size.value) {
		case BoSize.extra_small:
		case BoSize.small:
			return 'text-xs';
		case BoSize.large:
			return 'text-base';
		case BoSize.extra_large:
			return 'text-lg';
		case BoSize.default:
		default:
			return 'text-sm';
	}
});

const textClasses = computed<string>(() => {
	let classes = '';

	if (type.value === BoButtonType.outline) {
		switch (variant.value) {
			case BoVariant.danger:
				classes = TailwindUtils.merge(classes, 'text-red-600 hover:text-white');
				break;
			case BoVariant.warning:
				classes = TailwindUtils.merge(
					classes,
					'text-yellow-500 hover:text-white',
				);
				break;
			case BoVariant.success:
				classes = TailwindUtils.merge(
					classes,
					'text-green-500 hover:text-white',
				);
				break;
			case BoVariant.dark:
				classes = TailwindUtils.merge(
					classes,
					'text-gray-500 hover:text-white dark:hover:text-white',
				);
				break;
			case BoVariant.purple:
				classes = TailwindUtils.merge(
					classes,
					'text-purple-500 hover:text-white',
				);
				break;
			case BoVariant.teal:
				classes = TailwindUtils.merge(
					classes,
					'text-teal-500 hover:text-white',
				);
				break;
			case BoVariant.secondary:
				classes = TailwindUtils.merge(
					classes,
					'text-gray-700 hover:text-white dark:text-gray-300 dark:hover:text-white',
				);
				break;
			case BoVariant.primary:
			case BoVariant.default:
			default:
				classes = TailwindUtils.merge(
					classes,
					'text-blue-700 hover:text-white dark:text-blue-500 dark:hover:text-white',
				);
				break;
		}
	}

	if (type.value === BoButtonType.pill || type.value === BoButtonType.default) {
		switch (variant.value) {
			case BoVariant.secondary:
				classes = TailwindUtils.merge(classes, 'text-gray-600');
				break;
			case BoVariant.danger:
			case BoVariant.warning:
			case BoVariant.success:
			case BoVariant.dark:
			case BoVariant.purple:
			case BoVariant.teal:
			case BoVariant.primary:
			default:
				classes = TailwindUtils.merge(classes, 'text-white');
				break;
		}
	}

	classes = TailwindUtils.merge(classes, 'font-medium');
	return classes;
});

const typeClasses = computed(() => {
	switch (type.value) {
		case BoButtonType.pill:
			return 'rounded-full';
		case BoButtonType.outline:
		case BoButtonType.default:
		default:
			return 'rounded-lg';
	}
});

const sizeClasses = computed<string>(() => {
	switch (size.value) {
		case BoSize.extra_small:
			return 'px-2 py-1';
		case BoSize.small:
			return 'px-2 py-1.5';
		case BoSize.large:
			return 'px-5 py-2.5';
		case BoSize.extra_large:
			return 'px-6 py-3';
		case BoSize.default:
		default:
			return 'px-3 py-2';
	}
});

const colorClasses = computed<string>(() => {
	switch (type.value) {
		case BoButtonType.outline:
			switch (variant.value) {
				case BoVariant.danger:
					return 'border border-red-600 hover:bg-red-600 focus:ring-red-600';
				case BoVariant.warning:
					return 'border border-yellow-500 hover:bg-yellow-600 focus:ring-yellow-500';
				case BoVariant.success:
					return 'border border-green-500 hover:bg-green-600 focus:ring-green-500';
				case BoVariant.dark:
					return 'border border-gray-800 hover:bg-gray-700 focus:ring-gray-800';
				case BoVariant.purple:
					return 'border border-purple-500 hover:bg-purple-600 focus:ring-purple-500';
				case BoVariant.teal:
					return 'border border-teal-500 hover:bg-teal-600 focus:ring-teal-500';
				case BoVariant.secondary:
					return 'border border-gray-500 hover:bg-gray-600 focus:ring-gray-500';
				case BoVariant.primary:
				default:
					return 'border border-blue-700 hover:bg-blue-800 focus:ring-blue-700';
			}
		case BoButtonType.pill:
		case BoButtonType.default:
		default:
			switch (variant.value) {
				case BoVariant.danger:
					return 'bg-red-600 hover:bg-red-700 focus:ring-red-600';
				case BoVariant.warning:
					return 'bg-yellow-500 hover:bg-yellow-600 focus:ring-yellow-500';
				case BoVariant.success:
					return 'bg-green-600 hover:bg-green-700 focus:ring-green-600';
				case BoVariant.dark:
					return 'bg-gray-800 hover:bg-gray-700 focus:ring-gray-800';
				case BoVariant.secondary:
					return 'bg-gray-200 hover:bg-gray-300 focus:ring-gray-200';
				case BoVariant.purple:
					return 'bg-purple-500 hover:bg-purple-600 focus:ring-purple-500';
				case BoVariant.teal:
					return 'bg-teal-500 hover:bg-teal-600 focus:ring-teal-500';
				case BoVariant.primary:
				default:
					return 'bg-blue-700 hover:bg-blue-800 focus:ring-blue-700';
			}
	}
});

const style = computed(() => {
	return TailwindUtils.merge(
		typeClasses.value,
		sizeClasses.value,
		colorClasses.value,
		disabledClasses,
		textClasses.value,
		textSizes.value,
	);
});
</script>
