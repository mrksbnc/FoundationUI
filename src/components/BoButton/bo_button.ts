import {
	type BoButtonClassMap,
	type UseBoButtonCompoableArgs,
} from '@/components/BoButton';
import { BoSize } from '@/constants';
import { TailwindUtils } from '@/utils';
import { computed } from 'vue';

export enum BoButtonType {
	default = 'default',
	pill = 'pill',
	outline = 'outline',
}

export enum BoButtonVariant {
	primary = 'primary',
	secondary = 'secondary',
	success = 'success',
	warning = 'warning',
	danger = 'danger',
	dark = 'dark',
	purple = 'purple',
	teal = 'teal',
}

const BO_BUTTON_TRANSITION_CLASSES =
	/*tw*/ 'transition-all duration-200 ease-in-out' as const;

const BO_BUTTON_FLEX_CLASSES =
	/*tw*/ 'flex items-center justify-center content-center text-center w-full max-w-fit space-x-1' as const;

const BO_BUTTON_DISABLED_CLASSES =
	/*tw*/ 'disabled:cursor-not-allowed disabled:opacity-50' as const;

const BO_BUTTON_DEFAULT_CLASSES: BoButtonClassMap<BoButtonVariant> =
	Object.freeze({
		default: {
			primary: /*tw*/ 'text-white dark:text-white bg-blue-700 dark:bg-blue-600',
			secondary:
				/*tw*/ 'text-gray-600 dark:text-gray-500 bg-gray-200 dark:bg-gray-100',
			success:
				/*tw*/ 'text-white dark:text-white bg-green-600 dark:bg-green-500',
			warning:
				/*tw*/ 'text-white dark:text-white bg-yellow-600 dark:bg-yellow-500',
			danger: /*tw*/ 'text-white dark:text-white bg-red-600 dark:bg-red-500',
			dark: /*tw*/ 'text-white dark:text-white bg-gray-600 dark:bg-gray-500',
			purple:
				/*tw*/ 'text-white dark:text-white bg-purple-600 dark:bg-purple-500',
			teal: /*tw*/ 'text-white dark:text-white bg-teal-600 dark:bg-teal-500',
		},
		active: {
			primary:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-blue-800 dark:hover:bg-blue-700',
			secondary:
				/*tw*/ 'hover:text-gray-800 dark:hover:text-gray-300 hover:bg-gray-300 dark:hover:bg-gray-200',
			success:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-green-700 dark:hover:bg-green-600',
			warning:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-yellow-700 dark:hover:bg-yellow-600',
			danger:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-red-700 dark:hover:bg-red-600',
			dark: /*tw*/ 'hover:text-gray-600 dark:hover:text-white hover:bg-gray-700 dark:hover:bg-gray-600',
			purple:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-purple-700 dark:hover:bg-purple-600',
			teal: /*tw*/ 'hover:text-white dark:hover:text-white hover:bg-teal-700 dark:hover:bg-teal-600',
		},
	});

const BO_BUTTON_OUTLINE_CLASSES: BoButtonClassMap<BoButtonVariant> =
	Object.freeze({
		default: {
			primary:
				/*tw*/ 'border border-blue-700 dark:border-blue-600 text-blue-700 dark:text-blue-600 bg-transparent',
			secondary:
				/*tw*/ 'border border-gray-700 dark:border-gray-600 text-gray-700 dark:text-gray-600 bg-transparent',
			success:
				/*tw*/ 'border border-green-600 dark:border-green-500 text-green-600 dark:text-green-500 bg-transparent',
			warning:
				/*tw*/ 'border border-yellow-600 dark:border-yellow-500 text-yellow-600 dark:text-yellow-500 bg-transparent',
			danger:
				/*tw*/ 'border border-red-600 dark:border-red-500 text-red-600 dark:text-red-500 bg-transparent',
			dark: /*tw*/ 'border border-gray-600 dark:border-gray-500 text-gray-600 dark:text-gray-500 bg-transparent',
			purple:
				/*tw*/ 'border border-purple-600 dark:border-purple-500 text-purple-600 dark:text-purple-500 bg-transparent',
			teal: /*tw*/ 'border border-teal-600 dark:border-teal-500 text-teal-600 dark:text-teal-500 bg-transparent',
		},
		active: {
			primary:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-blue-700 dark:hover:bg-blue-600',
			secondary:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-gray-500 dark:hover:bg-gray-500 hover:border-gray-500 dark:hover:border-gray-500',
			success:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-green-600 dark:hover:bg-green-500',
			warning:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-yellow-600 dark:hover:bg-yellow-500',
			danger:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-red-600 dark:hover:bg-red-500',
			dark: /*tw*/ 'hover:text-white dark:hover:text-white hover:bg-gray-600 dark:hover:bg-gray-500',
			purple:
				/*tw*/ 'hover:text-white dark:hover:text-white hover:bg-purple-600 dark:hover:bg-purple-500',
			teal: /*tw*/ 'hover:text-white dark:hover:text-white hover:bg-teal-600 dark:hover:bg-teal-500',
		},
	});

enum BoButtonBorderRadiusClasses {
	pill = /*tw*/ 'rounded-full',
	default = /*tw*/ 'rounded-lg',
	outline = /*tw*/ 'rounded-lg',
}

enum BoButtonPaddingClasses {
	extra_small = /*tw*/ 'px-2 py-1',
	small = /*tw*/ 'px-2 py-1.5',
	default = /*tw*/ 'px-3 py-2',
	large = /*tw*/ 'px-4 py-3',
	extra_large = /*tw*/ 'px-5 py-4',
}

enum BoButtonTextSizeClasses {
	extra_small = /*tw*/ 'text-xs',
	small = /*tw*/ 'text-xs',
	default = /*tw*/ 'text-sm',
	large = /*tw*/ 'text-base',
	extra_large = /*tw*/ 'text-lg',
}

enum BoButtonShadowClasses {
	blue = /*tw*/ 'shadow-sm shadow-blue-500/50 dark:shadow-sm dark:shadow-blue-800/80',
	gray = /*tw*/ 'shadow-sm shadow-gray-500/50 dark:shadow-sm dark:shadow-gray-800/80',
	red = /*tw*/ 'shadow-sm shadow-red-500/50 dark:shadow-sm dark:shadow-red-800/80',
	yellow = /*tw*/ 'shadow-sm shadow-yellow-500/50 dark:shadow-sm dark:shadow-yellow-800/80',
	green = /*tw*/ 'shadow-sm shadow-green-500/50 dark:shadow-sm dark:shadow-green-800/80',
	black = /*tw*/ 'shadow-sm shadow-black-500/50 dark:shadow-sm dark:shadow-black-800/80',
	purple = /*tw*/ 'shadow-sm shadow-purple-500/50 dark:shadow-sm dark:shadow-purple-800/80',
	teal = /*tw*/ 'shadow-sm shadow-teal-500/50 dark:shadow-sm dark:shadow-teal-800/80',
}

export const useBoButtonStyle = (
	props: UseBoButtonCompoableArgs,
): { classes: string; iconSize: BoSize } => {
	const variant = props.variant.value;
	const type = props.type.value;
	const size = props.size.value;

	const isOutline = computed<boolean>(() => type === BoButtonType.outline);

	const base = computed<string>(() => {
		let shadow = '';

		switch (variant) {
			case BoButtonVariant.primary:
				shadow = BoButtonShadowClasses.blue;
				break;
			case BoButtonVariant.secondary:
				shadow = BoButtonShadowClasses.gray;
				break;
			case BoButtonVariant.success:
				shadow = BoButtonShadowClasses.green;
				break;
			case BoButtonVariant.warning:
				shadow = BoButtonShadowClasses.yellow;
				break;
			case BoButtonVariant.danger:
				shadow = BoButtonShadowClasses.red;
				break;
			case BoButtonVariant.dark:
				shadow = BoButtonShadowClasses.black;
				break;
			case BoButtonVariant.purple:
				shadow = BoButtonShadowClasses.purple;
				break;
			case BoButtonVariant.teal:
				shadow = BoButtonShadowClasses.teal;
				break;
		}

		return isOutline.value
			? TailwindUtils.merge(
					BO_BUTTON_FLEX_CLASSES,
					BO_BUTTON_DISABLED_CLASSES,
					BO_BUTTON_TRANSITION_CLASSES,
					BO_BUTTON_OUTLINE_CLASSES.default[variant],
					BO_BUTTON_OUTLINE_CLASSES.active != null
						? BO_BUTTON_OUTLINE_CLASSES.active[variant]
						: '',
				)
			: TailwindUtils.merge(
					shadow,
					BO_BUTTON_FLEX_CLASSES,
					BO_BUTTON_DISABLED_CLASSES,
					BO_BUTTON_TRANSITION_CLASSES,
					BO_BUTTON_DEFAULT_CLASSES.default[variant],
					BO_BUTTON_DEFAULT_CLASSES.active != null
						? BO_BUTTON_DEFAULT_CLASSES.active[variant]
						: '',
				);
	});

	const borderRadius = computed<string>(() => {
		return BoButtonBorderRadiusClasses[type];
	});

	const buttonAndContentSize = computed<string>(() => {
		switch (size) {
			case BoSize.extra_small:
				return TailwindUtils.merge(
					BoButtonPaddingClasses.extra_small,
					BoButtonTextSizeClasses.extra_small,
				);
			case BoSize.small:
				return TailwindUtils.merge(
					BoButtonPaddingClasses.small,
					BoButtonTextSizeClasses.small,
				);
			case BoSize.large:
				return TailwindUtils.merge(
					BoButtonPaddingClasses.large,
					BoButtonTextSizeClasses.large,
				);
			case BoSize.extra_large:
				return TailwindUtils.merge(
					BoButtonPaddingClasses.extra_large,
					BoButtonTextSizeClasses.extra_large,
				);
			case BoSize.default:
			default:
				return TailwindUtils.merge(
					BoButtonPaddingClasses.default,
					BoButtonTextSizeClasses.default,
				);
		}
	});

	const classes = computed<string>(() => {
		return TailwindUtils.merge(
			base.value,
			borderRadius.value,
			buttonAndContentSize.value,
		);
	});

	const iconSize = computed<BoSize>(() => {
		switch (size) {
			case BoSize.large:
			case BoSize.extra_large:
				return BoSize.default;
			case BoSize.extra_small:
			case BoSize.small:
			case BoSize.default:
			default:
				return BoSize.small;
		}
	});

	return {
		classes: classes.value,
		iconSize: iconSize.value,
	};
};
