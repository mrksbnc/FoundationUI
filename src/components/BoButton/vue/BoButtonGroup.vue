import BoIcon from '@/components/BoIcon/vue/BoIcon.vue';
<template>
	<div
		role="group"
		class="inline-flex flex-row"
	>
		<button
			v-for="(item, index) in items"
			:key="item.id"
			type="button"
			role="button"
			:disabled="disabled"
			:class="buttonGroupClasses"
			class="inline-flex items-center space-x-2 border-b border-t shadow-md first:rounded-s-lg first:border last:rounded-e-lg last:border"
			@click="handleSelect(index, item)"
		>
			<bo-icon
				v-if="item.prefixIcon"
				:icon="item.prefixIcon"
				:size="iconSize"
			/>
			<span>{{ item.label }}</span>
			<bo-icon
				v-if="item.suffixIcon"
				:icon="item.suffixIcon"
				:size="iconSize"
			/>
		</button>
	</div>
</template>

<script setup lang="ts">
import { BoIcon } from '@/components/BoIcon';
import {
	BoButtonGroupSize,
	BoButtonGroupVariant,
	ButtonGroupEvent,
	type BoButtonGroupItem,
} from '@/components/BoButton';
import { computed, toRefs, type PropType } from 'vue';
import { BoSize } from '@/constants';
import { TailwindUtils } from '@/utils';

const emits = defineEmits<{
	(
		e: ButtonGroupEvent.SELECT,
		payload: { index: number; item: BoButtonGroupItem },
	): void;
}>();

const props = defineProps({
	items: {
		type: Array as PropType<BoButtonGroupItem[]>,
		required: true,
	},
	disabled: {
		type: Boolean,
		default: false,
	},
	size: {
		type: String as PropType<BoButtonGroupSize>,
		default: BoSize.default,
	},
	variant: {
		type: String as PropType<BoButtonGroupVariant>,
		default: BoButtonGroupVariant.default,
	},
});

const { items, disabled, size, variant } = toRefs(props);

const buttonGroupClasses = computed<string>(() => {
	let classes = '';

	switch (size.value) {
		case BoButtonGroupSize.small:
			classes = TailwindUtils.merge(classes, 'text-xs px-3 py-1');
			break;
		case BoButtonGroupSize.default:
			classes = TailwindUtils.merge(classes, 'text-sm px-4 py-2');
			break;
		case BoButtonGroupSize.large:
			classes = TailwindUtils.merge(classes, 'text-base px-6 py-3');
			break;
	}

	switch (disabled.value) {
		case true:
			classes = TailwindUtils.merge(
				classes,
				'disabled:cursor-not-allowed disabled:opacity-50',
			);
			break;
		case false:
		default:
			classes = TailwindUtils.merge(classes, 'cursor-pointer');
			break;
	}

	if (variant.value === BoButtonGroupVariant.default) {
		const colorClasses = 'text-gray-700 border-gray-300 bg-white';
		const hoverClasses = 'hover:bg-gray-200';
		const darkColorClasses =
			'dark:bg-gray-800 dark:border-gray-700 dark:text-gray-300 dark:hover:bg-gray-700';
		const focusClasses = 'focus:bg-gray-100 focus:dark:bg-gray-700';

		classes = TailwindUtils.merge(
			classes,
			colorClasses,
			hoverClasses,
			darkColorClasses,
			focusClasses,
		);
	} else if (variant.value === BoButtonGroupVariant.info) {
		const colorClasses = 'text-blue-600 border-blue-600';
		const hoverClasses = 'hover:bg-blue-600 hover:text-white';
		const darkColorClasses =
			'dark:border-blue-500 dark:text-blue-500 dark:hover:bg-blue-500 dark:hover:text-white';
		const focusClasses =
			'focus:text-white focus:dark:text-white focus:bg-blue-600 focus:dark:bg-blue-500';

		classes = TailwindUtils.merge(
			classes,
			colorClasses,
			hoverClasses,
			darkColorClasses,
			focusClasses,
		);
	} else if (variant.value === BoButtonGroupVariant.danger) {
		const colorClasses = 'text-red-600 border-red-600';
		const hoverClasses = 'hover:bg-red-600 hover:text-white';
		const darkColorClasses =
			'dark:border-red-500 dark:text-red-500 dark:hover:bg-red-500 dark:hover:text-white';
		const focusClasses =
			'focus:text-white focus:dark:text-white focus:bg-red-600 focus:dark:bg-red-500';

		classes = TailwindUtils.merge(
			classes,
			colorClasses,
			hoverClasses,
			darkColorClasses,
			focusClasses,
		);
	} else if (variant.value === BoButtonGroupVariant.success) {
		const colorClasses = 'text-green-600 border-green-600';
		const hoverClasses = 'hover:bg-green-600 hover:text-white';
		const darkColorClasses =
			'dark:border-green-500 dark:text-green-500 dark:hover:bg-green-500 dark:hover:text-white';
		const focusClasses =
			'focus:text-white focus:dark:text-white focus:bg-green-600 focus:dark:bg-green-500';

		classes = TailwindUtils.merge(
			classes,
			colorClasses,
			hoverClasses,
			darkColorClasses,
			focusClasses,
		);
	} else if (variant.value === BoButtonGroupVariant.warning) {
		const colorClasses = 'text-yellow-600 border-yellow-600';
		const hoverClasses = 'hover:bg-yellow-600 hover:text-white';
		const darkColorClasses =
			'dark:border-yellow-500 dark:text-yellow-500 dark:hover:bg-yellow-500 dark:hover:text-white';
		const focusClasses =
			'focus:text-white focus:dark:text-white focus:bg-yellow-600 focus:dark:bg-yellow-500';

		classes = TailwindUtils.merge(
			classes,
			colorClasses,
			hoverClasses,
			darkColorClasses,
			focusClasses,
		);
	} else if (variant.value === BoButtonGroupVariant.dark) {
		const colorClasses = 'text-gray-700 border-gray-700';
		const hoverClasses = 'hover:bg-gray-700 hover:text-white';
		const darkColorClasses =
			'dark:border-gray-300 dark:text-gray-300 dark:hover:bg-gray-300 dark:hover:text-white';
		const focusClasses =
			'focus:text-white focus:dark:text-white focus:bg-gray-700 focus:dark:bg-gray-300';

		classes = TailwindUtils.merge(
			classes,
			colorClasses,
			hoverClasses,
			darkColorClasses,
			focusClasses,
		);
	}

	return classes;
});

const iconSize = computed<BoSize>(() => {
	switch (size.value) {
		case BoButtonGroupSize.small:
			return BoSize.small;
		case BoButtonGroupSize.large:
			return BoSize.large;
		case BoButtonGroupSize.default:
		default:
			return BoSize.default;
	}
});

const handleSelect = (index: number, item: BoButtonGroupItem): void => {
	emits(ButtonGroupEvent.SELECT, { index, item });
};
</script>
