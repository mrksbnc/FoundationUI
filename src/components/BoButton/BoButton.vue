<template>
	<button
		type="button"
		:disabled="disabled"
		:class="[`bo-button--${variant}`, style]"
	>
		<span class="bo-button__prefix">
			<BoIcon
				v-if="prefixIcon"
				:icon="prefixIcon"
			/>
		</span>
		<span class="bo-button__label">
			{{ label }}
		</span>
		<span class="bo-button__suffix">
			<BoIcon
				v-if="suffixIcon"
				:icon="suffixIcon"
			/>
		</span>
	</button>
</template>

<script setup lang="ts">
import { computed, toRefs, type PropType } from 'vue';
import { BoButtonType, BoButtonVariant } from './bo_button';
import { BoSize } from '@/enums/size';
import BoIcon from '../BoIcon/BoIcon.vue';

const props = defineProps({
	label: {
		type: String,
		required: true,
	},
	variant: {
		type: String as PropType<BoButtonVariant>,
		default: BoButtonVariant.primary,
	},
	type: {
		type: String as PropType<BoButtonType>,
		default: BoButtonType.default,
	},
	disabled: {
		type: Boolean,
		default: false,
	},
	size: {
		type: String as PropType<BoSize>,
		default: BoSize.default,
	},
	prefixIcon: {
		type: String,
		default: null,
	},
	suffixIcon: {
		type: String,
		default: null,
	},
});

const { label, variant, type, disabled, size, prefixIcon, suffixIcon } =
	toRefs(props);

const disabledClasses = 'disabled:cursor-not-allowed disabled:opacity-50';

const sizeClasses = computed(() => {
	switch (size.value) {
		case BoSize.extra_small:
			return 'px-3 py-2';
		case BoSize.small:
			return 'px-3 py-2';
		case BoSize.large:
			return 'px-5 py-3';
		case BoSize.extra_large:
			return 'px-6 py-3.5';
		case BoSize.default:
		default:
			return 'px-5 py-2.5';
	}
});

const colorClasses = computed(() => {
	switch (type.value) {
		case BoButtonType.outline:
			switch (variant.value) {
				case BoButtonVariant.danger:
					return 'border border-red-500 text-white hover:bg-red-600 focus:ring-red-500';
				case BoButtonVariant.warning:
					return 'border-yellow-500 text-white hover:bg-yellow-600 focus:ring-yellow-500';
				case BoButtonVariant.success:
					return 'bg-green-500 text-white hover:bg-green-600 focus:ring-green-500';
				case BoButtonVariant.dark:
					return 'bg-gray-800 text-white hover:bg-gray-900 focus:ring-gray-800';
				case BoButtonVariant.light:
					return 'bg-gray-200 text-gray-800 hover:bg-gray-300 focus:ring-gray-200';
				case BoButtonVariant.link:
					return 'text-blue-500 hover:text-blue-600 focus:ring-blue-500';
				case BoButtonVariant.alternative:
					return 'bg-gray-200 text-gray-800 hover:bg-gray-300 focus:ring-gray-200';
				case BoButtonVariant.secondary:
					return 'bg-gray-500 text-white hover:bg-gray-600 focus:ring-gray-500';
				case BoButtonVariant.primary:
				default:
					return 'bg-blue-500 text-white hover:bg-blue-600 focus:ring-blue-500';
			}
		case BoButtonType.pill:
		case BoButtonType.default:
		default:
			switch (variant.value) {
				case BoButtonVariant.danger:
					return 'bg-red-500 text-white hover:bg-red-600 focus:ring-red-500';
				case BoButtonVariant.warning:
					return 'bg-yellow-500 text-white hover:bg-yellow-600 focus:ring-yellow-500';
				case BoButtonVariant.success:
					return 'bg-green-500 text-white hover:bg-green-600 focus:ring-green-500';
				case BoButtonVariant.dark:
					return 'bg-gray-800 text-white hover:bg-gray-900 focus:ring-gray-800';
				case BoButtonVariant.light:
					return 'bg-gray-200 text-gray-800 hover:bg-gray-300 focus:ring-gray-200';
				case BoButtonVariant.link:
					return 'text-blue-500 hover:text-blue-600 focus:ring-blue-500';
				case BoButtonVariant.alternative:
					return 'bg-gray-200 text-gray-800 hover:bg-gray-300 focus:ring-gray-200';
				case BoButtonVariant.secondary:
					return 'bg-gray-500 text-white hover:bg-gray-600 focus:ring-gray-500';
				case BoButtonVariant.primary:
				default:
					return 'bg-blue-500 text-white hover:bg-blue-600 focus:ring-blue-500';
			}
	}
});

const style = computed(() => {
	return `${sizeClasses.value} ${colorClasses.value} ${disabled.value ? disabledClasses : ''}`;
});
</script>
