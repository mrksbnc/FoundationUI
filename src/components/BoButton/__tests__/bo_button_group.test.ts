import { BoButtonGroupSize, BoButtonGroupVariant } from '@/components/BoButton';
import BoButtonGroup from '@/components/BoButton/vue/BoButtonGroup.vue';
import { mount } from '@vue/test-utils';
import { describe, expect, it } from 'vitest';

describe('BoButtonGroup', () => {
	it('renders the correct number of buttons', () => {
		const items = [
			{ id: 1, label: 'Button 1' },
			{ id: 2, label: 'Button 2' },
			{ id: 3, label: 'Button 3' },
		];
		const wrapper = mount(BoButtonGroup, {
			props: {
				items,
			},
		});

		const buttons = wrapper.findAll('button');
		expect(buttons.length).toBe(items.length);
	});

	it('emits the correct event when a button is clicked', async () => {
		const items = [
			{ id: 1, label: 'Button 1' },
			{ id: 2, label: 'Button 2' },
			{ id: 3, label: 'Button 3' },
		];
		const wrapper = mount(BoButtonGroup, {
			props: {
				items,
			},
		});

		const buttons = wrapper.findAll('button');
		const buttonIndex = 1;
		await buttons[buttonIndex].trigger('click');

		expect(wrapper.emitted().select).toBeTruthy();
		expect(wrapper.emitted().select[0]).toEqual([
			{ index: buttonIndex, item: items[buttonIndex] },
		]);
	});

	it('applies the correct classes based on props', () => {
		const items = [
			{ id: 1, label: 'Button 1' },
			{ id: 2, label: 'Button 2' },
			{ id: 3, label: 'Button 3' },
		];
		const disabled = true;
		const size = BoButtonGroupSize.small;
		const variant = BoButtonGroupVariant.default;

		const wrapper = mount(BoButtonGroup, {
			props: {
				items,
				disabled,
				size,
				variant,
			},
		});

		const buttonGroup = wrapper.find('.inline-flex');
		expect(buttonGroup.classes()).toContain('flex-row');

		const buttons = wrapper.findAll('button');
		buttons.forEach((button) => {
			expect(button.classes()).toContain('disabled:cursor-not-allowed');
			expect(button.classes()).toContain('disabled:opacity-50');
			expect(button.classes()).toContain('text-xs');
			expect(button.classes()).toContain('text-gray-700');
		});
	});
});
