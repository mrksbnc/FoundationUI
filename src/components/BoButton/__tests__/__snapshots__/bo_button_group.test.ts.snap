// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BoButtonGroup > should match the snapshot 1`] = `
"<div role="group" class="inline-flex flex-row"><button type="button" role="button" class="text-sm px-4 py-2 cursor-pointer text-gray-700 border-gray-300 bg-white hover:bg-gray-200 dark:bg-gray-800 dark:border-gray-700 dark:text-gray-300 dark:hover:bg-gray-700 focus:bg-gray-100 focus:dark:bg-gray-700 inline-flex items-center first:border last:border border-t border-b first:rounded-s-lg last:rounded-e-lg shadow-md space-x-2">
    <!--v-if--><span>Button 1</span>
    <!--v-if-->
  </button><button type="button" role="button" class="text-sm px-4 py-2 cursor-pointer text-gray-700 border-gray-300 bg-white hover:bg-gray-200 dark:bg-gray-800 dark:border-gray-700 dark:text-gray-300 dark:hover:bg-gray-700 focus:bg-gray-100 focus:dark:bg-gray-700 inline-flex items-center first:border last:border border-t border-b first:rounded-s-lg last:rounded-e-lg shadow-md space-x-2">
    <!--v-if--><span>Button 2</span>
    <!--v-if-->
  </button><button type="button" role="button" class="text-sm px-4 py-2 cursor-pointer text-gray-700 border-gray-300 bg-white hover:bg-gray-200 dark:bg-gray-800 dark:border-gray-700 dark:text-gray-300 dark:hover:bg-gray-700 focus:bg-gray-100 focus:dark:bg-gray-700 inline-flex items-center first:border last:border border-t border-b first:rounded-s-lg last:rounded-e-lg shadow-md space-x-2">
    <!--v-if--><span>Button 3</span>
    <!--v-if-->
  </button></div>"
`;
